primitives/src/secondary_key.rs

```diff
@@ -13,32 +13,103 @@
+    iter,
 };
 
-// use crate::entity::IgnoredCaseString;
+/// Asset permissions.
+pub type AssetPermissions = SubsetRestriction<Ticker>;
+
+/// A permission to call
+///
+/// - specific functions, using `SubsetRestriction(Some(_))`, or
+///
+/// - all functions, using `SubsetRestriction(None)`
+///
+/// within a given pallet `pallet_name`.
+#[derive(Decode, Encode, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct PalletPermissions {
+    /// The name of a pallet.
+    pub pallet_name: PalletName,
+    /// A subset of function names within the pallet.
+    pub function_names: SubsetRestriction<FunctionName>,
+}
 
-/// Key permissions.
-/// # TODO
-/// 2. Review documents:
-///     - [MESH-235](https://polymath.atlassian.net/browse/MESH-235)
-///     - [Polymesh: Roles/Permissions](https://docs.google.com/document/d/12u-rMavow4fvidsFlLcLe7DAXuqWk8XUHOBV9kw05Z8/)
-#[allow(missing_docs)]
-#[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, Debug, PartialOrd, Ord)]
+impl PalletPermissions {
+    /// Constructs new pallet permissions from given arguments.
+    pub fn new(pallet_name: PalletName, function_names: SubsetRestriction<FunctionName>) -> Self {
+        PalletPermissions {
+            pallet_name,
+            function_names,
+        }
+    }
+
+    /// Constructs new pallet permissions for full access to pallet `pallet_name`.
+    pub fn entire_pallet(pallet_name: PalletName) -> Self {
+        PalletPermissions {
+            pallet_name,
+            function_names: SubsetRestriction(None),
+        }
+    }
+}
+
+/// Extrinsic permissions.
+pub type ExtrinsicPermissions = SubsetRestriction<PalletPermissions>;
+
+/// Portfolio permissions.
+pub type PortfolioPermissions = SubsetRestriction<PortfolioNumber>;
+
+/// Signing key permissions.
+///
+/// Common cases of permissions:
+/// - `Permissions::empty()`: no permissions,
+/// - `Permissions::default()`: full permissions.
+#[derive(Encode, Decode, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-pub enum Permission {
-    Full,
-    Admin,
-    Operator,
-    SpendFunds,
-    Custom(u8),
+pub struct Permissions {
+    /// The subset of assets under management.
+    pub asset: AssetPermissions,
+    /// The subset of callable extrinsics.
+    pub extrinsic: ExtrinsicPermissions,
+    /// The subset of portfolios management.
+    pub portfolio: PortfolioPermissions,
+}
+
+impl Permissions {
+    /// The empty permissions.
+    pub fn empty() -> Self {
+        Self {
+            asset: SubsetRestriction::empty(),
+            extrinsic: SubsetRestriction::empty(),
+            portfolio: SubsetRestriction::empty(),
+        }
+    }
+
+    /// Empty permissions apart from given extrinsic permissions.
+    pub fn from_pallet_permissions<I>(pallet_permissions: I) -> Self
+    where
+        I: IntoIterator<Item = PalletPermissions>,
```

Comment:
General nit: here's a good place to use `impl IntoIterator<...>`, as type inference is driven by whatever you pass as an argument. (Applies to other places in the PR.)